File: src/core/user/dto/create-user.dto.ts
import {
  IsEmail,
  IsNotEmpty,
  IsString,
  IsStrongPassword,
} from 'class-validator';

export class CreateUserDto {
  @IsEmail({}, { message: 'Digite um email válido' })
  public email: string;

  @IsString({ message: 'Digite um nome válido' })
  @IsNotEmpty({ message: 'Digite um nome válido' })
  public name: string;

  @IsStrongPassword(
    {},
    {
      message:
        'Digite uma senha com ao menos 8 caracteres, incluindo letras maiúsculas, minúsculas, números e símbolos',
    },
  )
  public password: string;
}

File: src/main/java/dev/joaopdias/vox/core/user/dto/CreateUserDto.java
package dev.joaopdias.vox.core.user.dto;

import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;

public record CreateUserDto(
    @NotBlank(message = "O e-mail é obrigatório")
    @Email(message = "Informe um e-mail válido")
    @Size(max = 320, message = "O e-mail deve ter no máximo 320 caracteres")
    String email,

    @NotBlank(message = "O nome é obrigatório")
    @Size(min = 2, max = 150, message = "O nome deve ter entre 2 e 150 caracteres")
    String name,

    @NotBlank(message = "A senha é obrigatória")
    @Size(min = 8, max = 72, message = "A senha deve ter entre 8 e 72 caracteres")
    @Pattern(
        regexp = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&._#\\-])[A-Za-z\\d@$!%*?&._#\\-]+$",
        message = "A senha deve conter ao menos uma letra maiúscula, uma minúscula, um número e um caractere especial"
    )
    String password
) {
}